File: services/http-gateway/app/services/jwt_session.rb
# frozen_string_literal: true

require "jwt"

class JwtSession
  COOKIE_NAME = "sellbot_jwt"

  def initialize(secret: GatewayConfig.jwt_secret, ttl_hours: GatewayConfig.jwt_ttl_hours)
    @secret = secret
    @ttl_hours = ttl_hours
  end

  def encode(seller_id:, tg_user_id:)
    payload = {
      seller_id: seller_id,
      tg_user_id: tg_user_id,
      exp: @ttl_hours.hours.from_now.to_i
    }
    JWT.encode(payload, @secret, "HS256")
  end

  def decode(token)
    return nil if token.blank?

    payload, = JWT.decode(token, @secret, true, algorithm: "HS256")
    {
      seller_id: payload["seller_id"].to_i,
      tg_user_id: payload["tg_user_id"].to_i
    }
  rescue JWT::DecodeError
    nil
  end
end

File: services/core/src/main/kotlin/com/sellbot/core/http/auth/JwtSession.kt
package com.sellbot.core.http.auth

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.exceptions.JWTVerificationException
import org.springframework.stereotype.Component
import com.sellbot.core.http.SellbotHttpProperties
import java.time.Instant
import java.time.temporal.ChronoUnit

const val JWT_COOKIE_NAME = "sellbot_jwt"

data class JwtPayload(
    val sellerId: Long,
    val tgUserId: Long,
)

@Component
class JwtSession(
    properties: SellbotHttpProperties,
) {
    private val secret = properties.jwt.secret
    private val ttlHours = properties.jwt.ttlHours
    private val algorithm = Algorithm.HMAC256(secret)

    fun encode(sellerId: Long, tgUserId: Long): String =
        JWT.create()
            .withClaim("seller_id", sellerId)
            .withClaim("tg_user_id", tgUserId)
            .withExpiresAt(Instant.now().plus(ttlHours, ChronoUnit.HOURS))
            .sign(algorithm)

    fun decode(token: String?): JwtPayload? {
        if (token.isNullOrBlank()) return null
        return try {
            val decoded = JWT.require(algorithm).build().verify(token)
            val sellerId = decoded.getClaim("seller_id").asLong()
            val tgUserId = decoded.getClaim("tg_user_id").asLong()
            if (sellerId == null || sellerId == 0L) return null
            JwtPayload(sellerId = sellerId, tgUserId = tgUserId ?: 0L)
        } catch (_: JWTVerificationException) {
            null
        }
    }
}
